VectorTools::project for p::d::Triangulations

masterleinad opened the pull request:
Currently, `VectorTools::project_distributed` only works without `project_to_boundary_first` and without `enforce_zero_boundary`.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -74,6 +77,81 @@
+namespace
+{
+  template <int dim, int fe_degree, int components, typename number>
+  class MassOperator : public Subscriptor
```
i would move this to `matrix_free/operators.h` and add a unit test to compare with, say, Trilinos.

davydden replied:
you can also derive it from `Base` and add diagonal to have a better preconditioner. Not that it will matter much in this context, but still.

kronbichler replied:
I think that the diagonal - or lumped diagonal, which is much easier to compute than the diagonal because you get it by applying the mass matrix on a vector of ones - would indeed be useful. As soon as the meshes are not exactly uniform or polynomial degrees are higher than 2, the different scaling indeed helps iteration counts.

masterleinad replied:
Yes, sounds like a good idea!

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -952,12 +1137,12 @@ namespace VectorTools
 
 
-  template <int dim, int spacedim>
+  template <int dim, typename VectorType, int spacedim>
```
this is not backward compatible, but I think it's ok.

davydden commented on source/numerics/vector_tools_rhs.inst.in:
```diff
@@ -19,19 +19,34 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension :  SPACE_DIMENSIONS
+     Vector<double>                                              &);
+
+    template
+    void create_right_hand_side<deal_II_dimension,LinearAlgebra::distributed::Vector<double>,deal_II_space_dimension>
```
why no instantiation for PETSc and Trilinos?

davydden replied:
i guess you forgot to push to you branch.

kronbichler commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -754,6 +754,8 @@ namespace VectorTools
    VectorType                                         &u2);
 
+
+
```
We only put one empty line between declarations.

kronbichler commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -1858,41 +1900,41 @@ namespace VectorTools
                                const Quadrature<dim>           &q,
                                const Function<spacedim,double> &rhs,
-                               Vector<double>                  &rhs_vector);
+                               VectorType                      &rhs_vector);
```
Why do you insert the `VectorType` between the `dim` and `spacedim` template arguments? We typically put the integer templates in front.

masterleinad replied:
I don't really have a preference, but at least `VectorTools::point_difference`, `VectorTools::point_value`, `VectorTools::point_gradient` and `VectorTools::compute_mean_value` use the same format. Since the other `VectorTools::project_*` use the format you are proposing, I'll follow your suggestion.

kronbichler commented on tests/mpi/project_distributed_common.h:
```diff
@@ -0,0 +1,358 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2006 - 2015 by the deal.II authors
```
Please adjust years

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -800,6 +880,74 @@ namespace VectorTools
+                                 const Quadrature<dim-1>                                   &q_boundary,
+                                 const bool                                                 project_to_boundary_first)
+    {
+      const parallel::distributed::Triangulation<dim,spacedim> *parallel_tria =
```
this should work with `parallel::Triangulation`, i.e. both `shared` and `distributed` as it has all you need here: mpi communicator and `is_locally_owned()`.

davydden replied:
also please add a unit test which use `p::s::Tria`.

kronbichler replied:
@davydden We also need to make sure that MatrixFree with `p::s::Tria` works. I just added #3285.

davydden replied:
@kronbichler correct, i was only thinking about assembly part, not the matrix-free solver itself.

@masterleinad :  i would still cast to `p::Tria` but would wait with `p::s::Tria` unit tests until we known that MatrixFree works with it.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -929,6 +1112,8 @@ namespace VectorTools
                 const hp::QCollection<dim-1>               &q_boundary,
                 const bool                                  project_to_boundary_first)
   {
+    Assert((dynamic_cast<const parallel::distributed::Triangulation<dim,spacedim>* > (&(dof.get_triangulation()))==0),
```
i think this assert is out of place, it not a _distributed_ `project`, right? Whereas you already have an assert in `do_project_distributed`.

masterleinad replied:
Yes, but I guess it's a good idea to fail with something more descriptive than `Can't ask for DoF indices on artificial cells.` if you accidentally use `project` instead of `project_distributed`.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -800,6 +880,74 @@ namespace VectorTools
+                                 const ConstraintMatrix                                    &constraints,
+                                 const Quadrature<dim>                                     &quadrature,
+                                 const Function<spacedim, typename VectorType::value_type> &function,
+                                 VectorType                                                &vec_result,
```
please document whether `vec_result` should be fully distributed or ghosted or something else.

davydden commented on tests/mpi/project_distributed_q_01.cc:
```diff
@@ -0,0 +1,31 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2006 - 2015 by the deal.II authors
```
adjust years

davydden commented on tests/mpi/project_distributed_q_01.cc:
```diff
@@ -0,0 +1,31 @@
+template <int dim>
+void test ()
+{
+  test_no_hanging_nodes<dim, 2, 1> (FESystem<dim>(FE_Q<dim>(1), 2));
```
maybe add more unit tests which use other functions like `test_with_hanging_nodes()`?

masterleinad commented:
I will have a look at all of this today.

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +205,115 @@ namespace MatrixFreeOperators
+   * @author Daniel Arndt, 2016
+   */
+  template <int dim, int fe_degree, int n_components = 1, typename number = double>
+  class MassOperator : public Subscriptor
```
please, derive from the `Base` class. It's there for a reason :smile:

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +205,115 @@ namespace MatrixFreeOperators
+     * Constructor. Initializes the shape information from the ShapeInfo field
+     * in the class FEEval.
+     */
+    MassOperator (const MatrixFree<dim,double> &data_in);
```
double -> number

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +205,115 @@ namespace MatrixFreeOperators
+
+  /**
+   * This class implements the operation of the action of a mass matrix
+   * and additionally the action of a lumped mass matrix that can be used for
```
there is nothing on `lumped mass matrix` in the code, AFAICT.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -800,6 +805,79 @@ namespace VectorTools
+      inhomogeneities*=-1.;
+
+      create_right_hand_side (mapping, dof, quadrature, function, rhs);
+      mass_matrix.vmult_add(rhs, inhomogeneities);
```
neat way of taking care of inhomogeneous constraints :+1:

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -158,7 +158,7 @@ namespace MatrixFreeOperators
 
     /**
-     * Get read access to diagonal of this operator.
+     * Get read access to the inverse diagonal of this operator.
```
thans for fixing

kronbichler replied:
While looking at this code, I wonder whether this class should actually hold the vector for the diagonal entries. I would prefer if there is simply a function that computes it, say a pure virtual function `compute_matrix_diagonal(VectorType&) const`. I know that this disallows using the `el(i,i)` method. On the other hand, if we plan to introduce a `DiagonalMatrix` class as dicussed in #3223 that class should hold the vector in my opinion. But then we would hold the diagonal twice, which can only lead to confusion. Finally, the inverse diagonal does not logically belong to the operator. (It would need some adjustments to `PreconditionChebyshev` but I have planned that anyway.)

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -857,6 +859,46 @@ namespace VectorTools
+   * template parameters.
+   * Currently, the implementation only allows for @p enforce_zero_boundary
+   * and @p project_to_boundary_first to be <tt>false</tt>.
+   */
```
please state that `vec_result` should be fully distributed, if I understand it correctly.

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -717,6 +760,84 @@ namespace MatrixFreeOperators
+    const std::vector<unsigned int> constrained_dofs
+      = Base<dim, number>::data->get_constrained_dofs();
+    for (unsigned int i=0; i<constrained_dofs.size(); ++i)
+      Base<dim, number>::inverse_diagonal_entries(constrained_dofs[i]) = 1.;
```
Here, you need to access the entry via `inverse_diagonal_entries.local_element(constrained_dofs[i])`.

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -717,6 +760,84 @@ namespace MatrixFreeOperators
+    for (unsigned int i=0; i<constrained_dofs.size(); ++i)
+      Base<dim, number>::inverse_diagonal_entries(constrained_dofs[i]) = 1.;
+
+    Base<dim, number>::inverse_diagonal_entries.compress(VectorOperation::insert);
```
Since you only access locally owned elements of the vector, this operation is not necessary as there should not be anything to be done. (And we only support `LinearAlgebra::distributed::Vector` here, so it does not make sense to use something else here.)

kronbichler commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -37,6 +38,9 @@
 #include <deal.II/lac/constraint_matrix.h>
+#include <deal.II/matrix_free/matrix_free.h>
+#include <deal.II/matrix_free/operators.h>
+#include <deal.II/matrix_free/fe_evaluation.h>
```
It is a pity that we need to include the *huge* file `fe_evaluation.h` inside the *huge* file `vector_tools.templates.h`. But we cannot really split it because we promise the user to let him/her include that single file in case an additional instantiation is needed. (Definitely not something for this PR to address.)

davydden commented on tests/mpi/project_distributed_common.h:
```diff
@@ -1,6 +1,6 @@
 // ---------------------------------------------------------------------
 //
-// Copyright (C) 2006 - 2015 by the deal.II authors
+// Copyright (C) 2016 - 2015 by the deal.II authors
```
2016 - 2015 is probably not what you wanted.

davydden commented on tests/mpi/project_distributed_q_01.cc:
```diff
@@ -1,6 +1,6 @@
 // ---------------------------------------------------------------------
 //
-// Copyright (C) 2006 - 2015 by the deal.II authors
+// Copyright (C) 2016 - 2016 by the deal.II authors
```
2016 is enough here.

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -1858,41 +1899,41 @@ namespace VectorTools
   void create_right_hand_side (const Mapping<dim, spacedim>    &mapping,
                                const DoFHandler<dim,spacedim>  &dof,
                                const Quadrature<dim>           &q,
                                const Function<spacedim,double> &rhs,
```
since you added new template, the `double -> typename VectorType::value_type`. Same everywhere else.

bangerth replied:
yes, good catch!

davydden commented on source/numerics/vector_tools_rhs.inst.in:
```diff
@@ -14,25 +14,24 @@
 
 
-for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension :  SPACE_DIMENSIONS)
+for (VEC : SERIAL_VECTORS; deal_II_dimension : DIMENSIONS; deal_II_space_dimension :  SPACE_DIMENSIONS)
```
unless you test that it compiles with complex-valued PETSc, probably it's better to instantiate with `REAL_SERIAL_VECTORS` instead.

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -717,6 +759,84 @@ namespace MatrixFreeOperators
+    Base<dim, number>()
+  {
+    Base<dim, number>::initialize(data);
+    compute_diagonal();
```
since this class would be used by users, do not call `compute_diagonal()` within a constructor. Users may not need diagonal. The may only need `vmult`.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -800,6 +806,85 @@ namespace VectorTools
+      additional_data.mpi_communicator = mpi_communicator;
+      additional_data.mapping_update_flags = (update_values | update_JxW_values);
+      MatrixFree<dim, number> matrix_free;
+      matrix_free.reinit (mapping, dof, constraints,
```
there is no `MatrixFree.reinit()` for `Mapping<dim,spacedim>` with `dim!=spacedim`. So i wonder if we should only have those functions for dim==spacedim?

davydden commented:
@masterleinad :   I checked out your PR to add those extra functions (already did that). I think you need to move function definitions into `.cc` and then add instantiations for `project_distributed()`.  Keeping definitions in `.templates.h` did not work for me with instantiations due to errors of 

```
template instantiation partial ordering for explicit instantiation <bla-bla> is ambiguous
```

The difference to other `project` is that we need to specify `fe_degree` and `n_components` in instantiations.

davydden commented:
for a record, here is my extension https://github.com/davydden/dealii/commit/8ca138291b17317e72710e19db28c8376b2d84fe

@bangerth Am I correct in that we would need to move definition of `project_distributed()` to `.cc` to have explicit instantiations with template arguments that can't be guessed from the input.

masterleinad commented:
@davydden Thanks for the really fast responses!

I would expect that including the `*.templates.h` should be sufficient.

davydden commented:
> I would expect that including the *.templates.h should be sufficient.

i would expect it too :smile:  but I could not make instantiation work. Please, try yourself for the code in this PR.

bangerth commented:
In C++, *where* you declare a function should not matter as long as the declaration is visible at the site of *use*. In other words, if it makes a difference whether you put the function into the `.templates.h` file or the `.cc` file, then something is wrong at a different level and it's worth trying to understand what it is.

davydden commented:
> In C++, where you declare a function should not matter as long as the declaration is visible at the site of use. In other words, if it makes a difference whether you put the function into the .templates.h file or the .cc file, then something is wrong at a different level and it's worth trying to understand what it is.

i replied here https://github.com/dealii/dealii/pull/3290

davydden commented on tests/mpi/project_distributed_common.h:
```diff
@@ -0,0 +1,358 @@
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/vector_tools.templates.h>
```
here is why you did not need to instantiate -- you include the `.templates.h`. My understanding is that this file is not be included in user code, everything has to be instantiated in `.inst.in` which is included from `.cc` file.

davydden replied:
in other words, you are still missing instantiation of `project_distributed()`, similar to what is done for `create_boundary_right_hand_side()`.

bangerth replied:
yes, correct. if you need the `#include` here, then you forgot to instantiate something somewhere.

davydden commented:
btw, do we really want to instantiate projection and friends for all vectors in `SERIAL_VECTORS`, which also include block vectors?  Would it make sense in this context? I would say it should be everything non-block.

davydden commented:
@masterleinad :  could you please address the remaining issues pointed out by myself and @kronbichler ? I really want to have this merged and use it in my PR to extend to what I need :smile:

masterleinad commented:
@davydden Yes, I will address the remaining issues today. 

At the moment I am struggling a bit with `constraints.compress(rhs)` when I have hanging nodes. `matrix_free.initialize_dof_vector` does not give access to all locally relevant dofs that are required in that function call.

davydden commented:
> At the moment I am struggling a bit with constraints.compress(rhs) when I have hanging nodes.

interesting, maybe one could extend `create_boundary_right_hand_side()` to take a pointer to `ConstraintMatrix` and if it is non-zero, do `constraints.distribute()`. Then it should work fine with fully distributed vector.

kronbichler commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -800,6 +806,85 @@ namespace VectorTools
+
+      create_right_hand_side (mapping, dof, quadrature, function, rhs);
+      mass_matrix.vmult_add(rhs, inhomogeneities);
+      constraints.condense(rhs);
```
@masterleinad Is it this condense call you're having trouble with? Do you really need this one? vmult_add should take care of the constraints internally because it calls distribute_local_to_global(). You might possibly want to call constraints.set_zero().

davydden replied:
I think this call actually should be right after `create_right_hand_side()`, which knows nothing about constraints and it's this RHS this needs condense. I think one needs to use `onstraints.distribute()` inside `create_right_hand_side()`.

masterleinad replied:
@kronbichler Yes, this is certainly true for the `inhomogeneities` part, but I think not for the part coming from `create_right_hand_side`.
If I just call `constraints.set_zero()`, I get wrong results for hanging nodes.

davydden commented:
@masterleinad may I suggest that you create a separate PR for `MassOperator` plus a unit test (e.g. compare with `SparseMatrix<double>`). This should be more-or-less ready, as far as I can tell. When we merge that, your projection PR and mine will be independent and we would be able to work on them in parallel. It would also make the reviewing easier as projection and definition of matrix-free mass operators are two separate things.

davydden commented:
ps. 

@masterleinad you don't need to remove mass matrix from here, just do `git checkout -b mass_matrix_mf`, `git reset master` and then commit only new matrix-free mass operator, followed by `git push...` and a separate PR. After that one is merged, you can just do `git merge master` to the current PR.

masterleinad commented:
@davydden Yes, I'll split this up.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -21,6 +21,7 @@
 #include <deal.II/base/function.h>
 #include <deal.II/base/quadrature.h>
 #include <deal.II/base/qprojector.h>
+#include <deal.II/distributed/tria.h>
```
it's enough to have `tria_base.h` here.

davydden commented:
Please rename functions to `project_parallel` to better reflect that in principle it should work both for `p::s::Tria` as well as `p::d::Tria`.

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -940,18 +937,6 @@ class PreconditionChebyshev : public Subscriptor
-   * Perform one step of the preconditioned Richardson iteration.
-   */
-  void step (VectorType       &dst,
-             const VectorType &src) const;
```
I think this is a change that's probably unrelated to what you want to achieve with the current patch (extend `VectorTools::project`). Could this be separated into a different pull request? (I'm asking because I don't know why you want to remove the function...)

masterleinad replied:
This is an artifact of rebasing. Thanks for pointing out.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -857,6 +857,47 @@ namespace VectorTools
+   * and @p project_to_boundary_first to be <tt>false</tt>.
+   * @p vec_result is assumed to not have a ghost elements.
+   */
+  template <int dim, typename VectorType, int spacedim, int components, int fe_degree>
```
We could have the same discussion here as I had in the other PR about `components` and `fe_degree` being part of the public interface, which I'm not totally excited about.

If they need to be part of the public interface, maybe sort them to the front so that one only has to specify those template arguments that are not automatically deducible by the compiler?

kronbichler replied:
I think we should remove both template arguments and provide pre-compiled paths. Regarding the number of components, if we pre-compile up to `4` components (3+1 for 3D), the rest can be treated with generic code in FEEvaluation that is upcoming.

masterleinad replied:
You mean kind of a `switch` for `components` and `fe_degree`?

davydden replied:
FYI, i did that here already https://github.com/dealii/dealii/pull/3262

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -895,6 +1021,12 @@ namespace VectorTools
+    Assert((dynamic_cast<const parallel::shared::Triangulation<dim,spacedim>* > (&(dof.get_triangulation()))==0),
+           ExcMessage("Use project_distributed() for parallel Triangulations!"));
+    Assert((dynamic_cast<const parallel::distributed::Triangulation<dim,spacedim>* > (&(dof.get_triangulation()))==0),
+           ExcMessage("Use project_distributed() for parallel Triangulations!"));
```
You can unify these two asserts if you do the `dynamic_cast` to `parallel::Triangulation`. That also takes care of possible future triangulation classes that happen to be parallel.

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -929,6 +1061,11 @@ namespace VectorTools
+    Assert((dynamic_cast<const parallel::shared::Triangulation<dim,spacedim>* > (&(dof.get_triangulation()))==0),
+           ExcNotImplemented());
+    Assert((dynamic_cast<const parallel::distributed::Triangulation<dim,spacedim>* > (&(dof.get_triangulation()))==0),
+           ExcNotImplemented());
```
same here and possible elsewhere below.

davydden commented:
i think it's a good idea to hid extra templates completely, because if someone gives a FE which is not supported by Matrix-free, we may in future switch to matrix-based projection.

masterleinad commented:
I guess the only thing left is hiding the template parameters.

masterleinad commented:
I am (hopefully) done with addressing the comments so far. In particular, the template arguments `fe_degree` and `components` are hidden and we explicitly instantiate `project_parallel_generic` for `fe_degree<9` and `components<5`.

`VectorTools::project_parallel` still only works without `project_to_boundary_first` and without `enforce_zero_boundary`. Therefore, it cannot replace `VectorTools::project`completely (yet). But the tests `numerics/project_q*` also pass with this implementation.

In contrast to #3290, the tests don't take #3294 into account.

davydden commented:
I think @bangerth is right -- all we do should work OOB for serial case. Entertain the idea of keeping old functions but switching to matrix-free if 1) FE_Q 2) extra flags are false. 3) optionally, if you encounter parallel tria and therefore can set MPI communicator to MatrixFree data object.

masterleinad commented:
> I think @bangerth is right -- all we do should work OOB for serial case. Entertain the idea of keeping old functions but switching to matrix-free if 1) FE_Q 2) extra flags are false. 3) optionally, if you encounter parallel tria and therefore can set MPI communicator to MatrixFree data object.

Yes, it works for the serial case as well. So is `project_matrix_free` a better name than `project_parallel` for our new functions?

I am fine with using the new function instead of `VectorTools::project` in that particular situation you describe.

davydden commented:
> project_matrix_free a better name than project_parallel for our new functions?

in my PR i kept the old name `project()`. My understanding is that @bangerth and @kronbichler are both in favour of this approach. Let's wait for the other guys to comment...

kronbichler commented:
Indeed, I would suggest that we only put `project` into the header file. Inside `vector_tools.templates.h`, I would suggest to continue using `project_parallel`, for conformity with the other interfaces. We might eventually create non-matrix-free paths that work in parallel, too, but let's face the problem when it comes up.

davydden commented:
@masterleinad could you please resolve the merge conflicts?

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -24,8 +24,14 @@
 #include <deal.II/hp/mapping_collection.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/matrix_free/matrix_free.h>
```
i dont think you need any of those added headers in `vector_tools.h`.

masterleinad replied:
Originally, this was needed for the implementation of `project_generic` in `vector_tools.h`. Since we know at least some use cases, it should be fine to move it to `vector_tools.templates.h` which is done now.

davydden replied:
but why does Github still shows this modifications? Did you push?

masterleinad replied:
No, I should have been more precise: ...which I am doing right now. Compiling really takes forever :unamused: (as noted in #3388).

davydden replied:
forever -- is a word dedicated to compiling Trilinos :smile:

masterleinad replied:
Here you go.

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -924,6 +930,64 @@ namespace VectorTools
+   * @p vec_result is assumed to not have a ghost elements.
+   */
+  template <int dim, typename VectorType, int spacedim>
+  void project_parallel (const Mapping<dim, spacedim>   &mapping,
```
i think the idea of @bangerth and @kronbichler is not to have any `project_parallel` in `vector_tools.h`. You only add them in internal namespace of `vector_tools.templates.h` to do the switching.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -38,9 +38,6 @@
 #include <deal.II/lac/vector_memory.h>
 #include <deal.II/lac/filtered_matrix.h>
 #include <deal.II/lac/constraint_matrix.h>
-#include <deal.II/matrix_free/matrix_free.h>
```
these would stay here.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -988,13 +1063,13 @@ namespace VectorTools
 
 
-    template <int dim, typename VectorType, int fe_degree, int n_q_points_1d>
+    template <int dim, typename VectorType, int spacedim, int fe_degree, int n_q_points_1d>
```
that's already fixed if you rebase (or merge `master` in).

masterleinad commented:
I will rebase again when replacing VectorTools::project works as expected.
Is there anything else missing?

davydden commented:
@masterleinad instead of rebasing you can just merge the current `master` to your branch, it could be easier to resolve such conflicts. As for the history, I would not bother as you would squash everything into a few commits anyway at the end.

davydden commented:
the reply below by @kronbichler is also relevant for this PR

> And one more remark: I recommend against using PreconditionChebyshev for solving a single system - at least in the mass matrix context. It needs to estimate eigenvalues by solving an auxiliary linear system. Until you have estimated the eigenvalues of the mass matrix, you have almost solved the system already. My suggestion is to use a Jacobi preconditioner

masterleinad commented:
The matrix-free implementation replaces `VectorTools::project` where possible and all of the `project` tests (`ctest -R "parallel`) are passing.
From my side, this is now ready and I will rebase and squash.

masterleinad commented:
@davydden Thanks for pointing me to @kronbichler's comment. `PreconditionJacobi` is now used instead.

davydden commented:
@masterleinad could you please merge in master, address two comments of myself above and then I and, maybe, somebody else will go through this PR again.

davydden commented on doc/news/changes.h:
```diff
@@ -434,6 +441,12 @@ inconvenience this causes.
  (Denis Davydov, 2016/10/31)
  </li>
 
+ <li> New: VectorTools::project_parallel extends VectorTools to 
```
there is no `project_parallel` now, please reword

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -958,7 +958,7 @@ namespace MatrixFreeOperators
                     const std::pair<unsigned int,unsigned int>  &cell_range) const
   {
-    FEEvaluation<dim, fe_degree, fe_degree+1, n_components, Number> phi(data);
+    FEEvaluation<dim, fe_degree, fe_degree+2, n_components, Number> phi(data);
```
i suppose you needed this to get extra precision in case of non-Q1 mapping to make unit tests pass?

masterleinad replied:
Correct, VectorTools::project is used with QGauss<dim>(fe.degree+2) in all the (previous) tests and the examples. This makes a difference for `numerics/project_*_03` and `tests/project_01_curved_boundary`.

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -924,6 +928,26 @@ namespace VectorTools
+   * @p vec_result is expected to not have any ghost entries.
+   */
+  template <int components, int fe_degree, int dim, typename VectorType, int spacedim>
+  void project_generic (const Mapping<dim, spacedim>                              &mapping,
```
do we want to have this function? I thought interface of `VectorTools` does not change at all, with a few exceptions of RHS calculations. So my understanding is that this should go to `.templates.h`.

davydden replied:
@bangerth @kronbichler  any opinions on this new function `project_generic` ?

davydden replied:
p.s. this function is used below in `project_parallel_helper`.

kronbichler replied:
I'm not sure how much we should expose. If we want a user to be able to use any of the 'unsupported' combinations, we need to expose all parallel functions. My plan was to have a fall-back that does not use templates (at least for the polynomial degree and the number of quadrature points), which comes in with the larger matrix-free update.

kronbichler replied:
But even in that case we would still be left with the question whether we want users to be able to use the faster path, which requires to either look into `vector_tools.templates.h` or expose them in the header file. Since adding manual instantiations with `vector_tools.templates.h` is tricky (I don't know how many of our users are doing that), we might also just refer to the implementation in the documentation in the header file. I'm not sure what is best...

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -2069,12 +2097,12 @@ namespace VectorTools
+  void create_boundary_right_hand_side (const Mapping<dim,spacedim>        &mapping,
+                                        const DoFHandler<dim,spacedim>     &dof,
+                                        const Quadrature<dim-1>            &q,
+                                        const Function<spacedim,double>    &rhs,
```
typename VectorType::value_type

masterleinad replied:
No idea, where this got lost...

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -2084,12 +2112,12 @@ namespace VectorTools
   void create_boundary_right_hand_side
   (const DoFHandler<dim,spacedim>     &dof,
    const Quadrature<dim-1>            &q,
    const Function<spacedim,double>    &rhs,
```
typename VectorType::value_type

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -2098,13 +2126,13 @@ namespace VectorTools
   (const hp::MappingCollection<dim,spacedim> &mapping,
    const hp::DoFHandler<dim,spacedim>        &dof,
    const hp::QCollection<dim-1>              &q,
    const Function<spacedim,double>           &rhs,
```
typename VectorType::value_type

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -2115,12 +2143,12 @@ namespace VectorTools
   void create_boundary_right_hand_side
   (const hp::DoFHandler<dim,spacedim> &dof,
    const hp::QCollection<dim-1>       &q,
    const Function<spacedim,double>    &rhs,
```
typename VectorType::value_type

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -917,7 +1050,7 @@ namespace VectorTools
       MatrixFree<dim, Number> matrix_free;
       matrix_free.reinit (mapping, dof, constraints,
-                          QGauss<1>(fe_degree+1), additional_data);
+                          QGauss<1>(fe_degree+2), additional_data);
```
please, don't change this here. For projection from quadrature data, the number of quadrature points is up to the user. Even though the mass matrix may not be integrated exactly, in most cases you would use `fe_degree+1`. So I am not in favour of this change and below.

masterleinad replied:
See above (https://github.com/dealii/dealii/pull/3262/files/6141097333690dc2445211ce712a97172224b56a#diff-77183bf880358cc7866c2938080ac4b2).
So far `fe_degree+2` is used in all tests and examples.

davydden replied:
i does not matter where it used in other tests examples. that's not the point. The point is that when you project from quadrature points, the 99% of usage case if `fe_degree+1`. So this case should be supported OOB. Please, do not modify `project_parallel` with `std_cxx11::function` in this PR.

davydden replied:
I will create a separate PR which does what you want without changing the templates...

davydden replied:
for the record, this is the PR https://github.com/dealii/dealii/pull/3393

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1109,34 +1242,34 @@ namespace VectorTools
     const unsigned int fe_degree = matrix_free.get_dof_handler().get_fe().degree;
 
-    Assert (fe_degree+1 == n_q_points_1d,
+    Assert (fe_degree+2 == n_q_points_1d,
```
that is, without those changes...

davydden commented on tests/Testing/Temporary/LastTest.log:
```diff
@@ -1,3 +1,3 @@
-Start testing: Sep 11 08:44 PDT
```
why is this file in Git and why do you update it? `Temporary` kind suggests that it should not be under version control at all...

masterleinad replied:
I guess this can safely go.

bangerth replied:
Already done: #3428.

davydden commented on tests/numerics/project_parallel_qpmf_q_01.cc:
```diff
@@ -23,9 +23,9 @@
 template <int dim>
 void test ()
 {
-  test_no_hanging_nodes<1, 2,dim> (FE_Q<dim>(1), 1);
```
please keep those as is

davydden commented on tests/numerics/project_parallel_qpmf_q_02.cc:
```diff
@@ -22,9 +22,9 @@
 template <int dim>
 void test ()
 {
-  test_with_hanging_nodes<1, 2,dim> (FE_Q<dim>(1), 1);
```
same here.

davydden commented on tests/numerics/project_parallel_q_01.cc:
```diff
@@ -0,0 +1,31 @@
+
+// check that VectorTools::project works for Q elements correctly
+
+char logname[] = "output";
```
the output should always be into `output` file, so there is no need to have this variable in every tests which uses `project_parallel_common.h`.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1252,113 +1548,115 @@ namespace VectorTools
-                               Vector<double>                 &rhs_vector)
+                               const Function<spacedim, typename VectorType::value_type> &rhs_function,
+                               VectorType                     &rhs_vector,
+                               const ConstraintMatrix         &constraints)
```
this should be mentioned in `INCOMPATABILITIES` because you changed `create_right_hand_side()` functions. I doubt anyone was using it, but nonetheless it was public within `VectorTools`.

masterleinad replied:
There is 
``` 
template <int dim, int spacedim, typename VectorType>
void create_right_hand_side (const Mapping<dim, spacedim>    &mapping,
                             const DoFHandler<dim,spacedim>  &dof,
                             const Quadrature<dim>           &q,
                             const Function<spacedim,typename VectorType::value_type> &rhs,
                             VectorType                      &rhs_vector,
                             const ConstraintMatrix          &constraints=ConstraintMatrix());
```
in `vector_tools.h` and this is used without specifying `ConstraintMatrix` in the old implementation of `VectorTools::project`.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1205,6 +1494,9 @@ namespace VectorTools
                 const hp::QCollection<dim-1>               &q_boundary,
                 const bool                                  project_to_boundary_first)
   {
+    Assert((dynamic_cast<const parallel::Triangulation<dim,spacedim>* > (&(dof.get_triangulation()))==0),
```
why is it not implemented? Even if you do `MatrixFree`, you don't have to have a parallel Triangulation. You just skip setting `mpi_communicator`. I think this Assert should go.

davydden replied:
right, i missed that it is `hp` version.

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -763,22 +763,25 @@ namespace VectorTools
-   * quadrature using the given quadrature rule; you should therefore make
-   * sure that the given quadrature formula is also sufficient for the
-   * integration of the mass matrix.
+   * This function performs numerical quadrature using the given quadrature rule;
```
is it still the case? I though you integrate mass matrix with `fe_degree+2` whereas the RHS can be done using user-provided quadrature formula.

masterleinad replied:
Correct.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1160,6 +1293,144 @@ namespace VectorTools
+      MatrixFree<dim,number>::AdditionalData::partition_color;
+    if (parallel_tria)
+      {
+        const MPI_Comm &mpi_communicator = parallel_tria->get_communicator();
```
you can assign the communicator straight away in one line, then you won't need brackets either.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1160,6 +1293,144 @@ namespace VectorTools
+    MatrixFree<dim, number> matrix_free;
+    matrix_free.reinit (mapping, dof, constraints,
+                        QGauss<1>(fe_degree+2), additional_data);
+    typedef MatrixFreeOperators::MassOperator<dim, fe_degree, components, number> MatrixType;
```
this will change to `MassOperator<dim, fe_degree, fe_degree+2, components, number>`

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1160,6 +1293,144 @@ namespace VectorTools
+    mass_matrix.vmult_add(rhs, inhomogeneities);
+
+    //now invert the matrix
+    ReductionControl     control(5*rhs.size(), 0., 1e-12, false, false);
```
i think @kronbichler pointed out in my PR that we don't really need five times the number of DoFs. Just loos the `5`.

masterleinad replied:
Then let me change this everywhere.

masterleinad replied:
Actually, the tests
- fe/fe_project_2d
- numerics/project_dgp_monomial_04
- numerics/project_dgp_monomial_05
- numerics/project_q_hierarchical_01
- numerics/project_q_hierarchical_02
- numerics/project_q_hierarchical_03
- numerics/project_q_hierarchical_04
- numerics/project_q_hierarchical_05

fail with `rhs.size()` and some also with `3.*rhs.size()`.

kronbichler replied:
That's quite bad. What is the polynomial degree in `FE_Q_Hierarchical`. I can imagine that happening if the polynomial degree gets four or higher because `FE_Q_Hierarchical` is really badly conditioned at higher degrees where roundoff forces more iterations.

In any case, I definitely agree that we should stay at `5*rhs.size()` then. It's not that the usual case with well-behaved mass matrices get slower...

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1160,6 +1293,144 @@ namespace VectorTools
+      use_matrix_free = false;
+    else
+      {
+        // Find out if the FiniteElement is supported
```
@kronbichler should this be moved into some static function of MatrixFree class like `static bool is_fe_supported(const FiniteElement<dim,spacedim> &fe)`?

kronbichler replied:
The check is currently done in `include/deal.II/matrix_free/shape_info.templates.h`, lines 79-113. I think it would indeed be a good idea to have a check in `MatrixFree` which could just use some logic in `shape_info.h`. I can implement that next week since I'm busy both today and tomorrow. If somebody beats me until then, I'm happy too.

bangerth replied:
what was the resolution about this?

masterleinad replied:
I would suggest to just open an issue for this and switch to the new check in `MatrixFree` once this is available.

masterleinad replied:
See #3614

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1171,13 +1442,31 @@ namespace VectorTools
-                project_to_boundary_first);
+    if (dim==spacedim)
+      {
+        const Mapping<dim> *const mapping_ptr = dynamic_cast<const Mapping<dim>*> (&mapping);
```
if `dim==spacedim`, should not the function below work without casting away the second (`spacedim`) template argument? I would expect it to work out-of-the-box.

masterleinad replied:
The problem is that we can't handle `dim!=spacedim` with the new implementation, but we instantiate also this case. The compiler complains that `Mapping<dim,spacedim>` can't be used in the matrix-free code despite the switch.

davydden replied:
right, presumably that's because it tries to compile this path even when dim!=spacedim.

bangerth replied:
Yes. Constant folding (e.g., the evaluation of `dim==spacedim`) and dead code removal happen at different times. We may get a facility for this in C++17: http://open-std.org/JTC1/SC22/WG21/docs/papers/2016/p0128r1.html

Until then, the idiom used here makes sense to me.

davydden commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -763,22 +763,26 @@ namespace VectorTools
    * that you need not give a second quadrature formula if you don't want to
    * project to the boundary first, but that you must if you want to do so.
+
+   * If the FiniteElement is supported by the MatrixFree class, neither
```
that's a bit log sentence which is hard to digest. Maybe organize those requirements into some bullet points?

davydden replied:
Or maybe write: "A matrix-free implementation is used when: (i)...(ii)... (iii).... Otherwise..."

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -917,8 +1050,8 @@ namespace VectorTools
       additional_data.mapping_update_flags = (update_values | update_JxW_values);
       MatrixFree<dim, Number> matrix_free;
       matrix_free.reinit (mapping, dof, constraints,
-                          QGauss<1>(fe_degree+2), additional_data);
```
why did you switch it back to `fe_degree+1`? Rebase errors? If so, `merge` is the answer :+1: Seriously.

masterleinad replied:
No, I thought you wanted me to leave the functions using `std_cxx11::function` as is, i.e. use `fe_degree+1`.
If I misunderstood you here, I happily make that consist with the original function again, i.e. use `fe_degree+2`.

davydden replied:
Exactly, i would prefer if you keep `std::function` implementations un-touched in this PR. In the current state of the `master` branch of deal.II this particular function uses `fe_degree+2`, so yes, as-is is what i wanted :smile:. Thanks!

davydden commented on tests/numerics/project_dgp_02.cc:
```diff
@@ -17,9 +17,6 @@
 
 // check that VectorTools::project works for DGP elements correctly
 
-char logname[] = "output";
```
i see now where that `char logname` comes from. I think this can indeed be removed, because we don't support comparing with files named other than `output`.

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -301,7 +301,7 @@ namespace MatrixFreeOperators
    *
    * @author Daniel Arndt, 2016
    */
-  template <int dim, int fe_degree, int n_components = 1, typename Number = double>
```
something is wrong on GitHub. This shall not be a part of the PR, as it is already done and that's how the `master` looks like now (with `n_q_points_1d`).

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -917,8 +1050,8 @@ namespace VectorTools
       additional_data.mapping_update_flags = (update_values | update_JxW_values);
       MatrixFree<dim, Number> matrix_free;
       matrix_free.reinit (mapping, dof, constraints,
-                          QGauss<1>(fe_degree+1), additional_data);
```
same here

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1004,7 +1137,7 @@ namespace VectorTools
       Assert (dof.get_fe().degree == fe_degree,
               ExcDimensionMismatch(fe_degree, dof.get_fe().degree));
 
-      typedef MatrixFreeOperators::MassOperator<dim, fe_degree, 1, Number> MatrixType;
```
and here -- this all should be in master already.

masterleinad commented:
Of course, I can avoid `merge` again. Rebasing with `master` is all I need. :smile:

davydden commented:
> Of course, I can avoid merge again. Rebasing with master is all I need.

anyway, once you cleanup the history, i hope GitHub won't show those changes. I have not idea why are they shown...

masterleinad commented:
Running the test suite, the tests `base/task_09` and `mpi/parallel_block_vector_03` fail, but this seems to be unrelated.

The two last commits modify tests in a way we probably don't want. While the output for `mapping_q_manifold` does not change, when passing from `MappingQ(..., true)` to  `MappingQ(..., false)`, the change in `fe_values_view_30` gives different results which are consistent with the previous old version of `VectorTools::project`.
Since this is not just an issue due to higher order mappings, I suspect that there are some problems with `MappingQ(...,true)` and `MappingQGeneric` or in general with mappings acting on internal faces.

masterleinad commented:
It seems that #3589 fixed the problem with the two failing tests `mapping_q_manifold_01` and `fe_values_view_30`. 
In particular, with #3593 we get consistent results between the old and new implementation in `mapping_q_manifold_01`. 
Similarly, changing the order of the quadrature rule in `fe_values_view_30` gives consistent results between the old and new implementation`.

That said, this PR should be good to go and I would appreciate someone to have another look.

bangerth commented:
There's of course again a merge conflict in the changes.h file. Can you rebase?

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -739,7 +739,6 @@ namespace VectorTools
                                  const ConstraintMatrix              &constraints,
```
You're changing the permissions of this file from 644 to 755. Can you undo this?

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -763,22 +762,31 @@ namespace VectorTools
-   * quadrature using the given quadrature rule; you should therefore make
-   * sure that the given quadrature formula is also sufficient for the
-   * integration of the mass matrix.
+
```
Add the missing asterisk on this line.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -763,22 +762,31 @@ namespace VectorTools
+   * - dim==spacedim
+   *
+   * In this case, this function performs numerical quadrature for integration
+   * of the provided function while a QGauss(fe_degree+2) object is used for
```
do you mean to say that "numerical quadrature USING THE GIVEN QUADRATURE FORMULA for integration..."?

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -763,22 +762,31 @@ namespace VectorTools
-   * sure that the given quadrature formula is also sufficient for the
-   * integration of the mass matrix.
+
+   * A MatrixFree implementation is used if the following consitions are met:
```
-> conditions

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -788,7 +796,8 @@ namespace VectorTools
    * @param[out] vec The output vector where the projected function will be
-   * stored in. This vector is required to be already initialized.
+   * stored in. This vector is required to be already initialized and must not
+   * be ghosted.
```
-> and must not have any ghost elements.

I think of the term "ghosted" to be colloquial.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -924,6 +933,29 @@ namespace VectorTools
                 VectorType &vec_result);
 
   /**
+   * Generic implementation for the project() function with FiniteElements that
```
-> finite elements

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -924,6 +933,29 @@ namespace VectorTools
+   * are supported by the MatrixFree class.
+   * Use this function if you use more than four components or the degree of
+   * your FiniteElement is higher than eight. You have to specify the first
+   * template arguments yourself.
```
Can you explain why this function should only be used if you have MORE components and a HIGHER polynomial degree?

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -924,6 +933,29 @@ namespace VectorTools
+   * implemented.
+   */
+  template <int components, int fe_degree, int dim, typename VectorType, int spacedim>
+  void project_generic (const Mapping<dim, spacedim>                              &mapping,
```
What makes this function "generic", as opposed to the other functions of same name? I don't think the name is particularly enlightening.

davydden replied:
but do we actually need that `project_generic`? why not do the switch internally in anonymous namespace in `templates.h`? I think we shall not confuse the user and provide `project()` which internally switches to matrix-free if it can and do something else otherwise.

bangerth replied:
That would be my preference as well.

masterleinad replied:
Just to clarify: We do this switch already for `project` in an anonymous namespace.
If I understand you correctly, you want me to just remove `project_generic` from public view and hence restrict the user to the instantiations in `project` until there is a fallback for `MatrixFree`?

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -19,6 +19,7 @@
 
```
same here with file permissions

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -80,6 +85,131 @@ DEAL_II_NAMESPACE_OPEN
 
 namespace VectorTools
 {
+  namespace implementation
```
We've typically called such namespaces `internal`, and changed the order of them so that `internal` is the outer namespace:
```
  namespace internal
  {
    namespace VectorTools
    {
       ...
```

bangerth replied:
To explain, this order has the effect of putting the documentation for these functions onto a separate doxygen page. There is no point in the doxygen page for `namespace VectorTools` to contain a link to `namespace VectorTools::internal` -- it's of no interest to users.

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -80,6 +85,131 @@ DEAL_II_NAMESPACE_OPEN
+          VectorTools::project_generic<components, 8>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
```
Didn't we limit the polynomial degree for which we provide instantiations in @davydden 's patch to four?

davydden replied:
no, we did not. The PR is here https://github.com/dealii/dealii/pull/3500, but there is no data to support the necessity of this change so far. So it's an open question at the moment.

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -968,7 +1101,7 @@ namespace VectorTools
 
       //now invert the matrix
-      ReductionControl     control(rhs.size(), 0., 1e-12, false, false);
+      ReductionControl     control(5.*rhs.size(), 0., 1e-12, false, false);
```
What was the reason again for this change? CG should surely converge in `rhs.size()` iterations.

kronbichler replied:
Roundoff errors in the conjugate gradient method in case there are bad meshes and/or bad shape functions. We have a bunch of such test cases. An example I recall uses `FE_Q_Hierarchical` at a degree four or five where conditioning gets so bad that we get into trouble. Now of course this indicates a delicate linear system to start with and just because the CG error estimate at some point beyond `rhs.size()` iterations is happy does not mean that the target tolerance is met - but I do not think that it should be addressed in this PR.

bangerth replied:
I see. Thanks for the explanation.

@masterleinad -- can you make @kronbichler 's remark into a (slightly shorter) comment that one could put in front of this line and the other, similar ones in this file? -- Thanks!

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1038,7 +1171,7 @@ namespace VectorTools
 
       //now invert the matrix
-      ReductionControl     control(rhs.size(), 0., 1e-12, false, false);
+      ReductionControl     control(5.*rhs.size(), 0., 1e-12, false, false);
```
same here

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1160,6 +1293,141 @@ namespace VectorTools
+    additional_data.tasks_parallel_scheme =
+      MatrixFree<dim,number>::AdditionalData::partition_color;
+    if (parallel_tria)
+      additional_data.mpi_communicator = parallel_tria->get_communicator();
```
since this is, I believe, the only place where you need `parallel_tria`, why not inline the dynamic cast into the condition right away?

masterleinad replied:
I need the pointer to set `mpi_communicator` in `additional_data`.

davydden replied:
i think what @bangerth means is that you inline the cast into the `if` and then do it again to get communicator on this line. End result is that you wouldn't need `parallel_tria` variable at all.

kronbichler replied:
The other question is whether it actually makes sense to be able to specify a separate communicator here. Given a mesh, the algorithm has no other choice than using the communicator of the mesh because it won't be able to get different ownership or locally owned ranges anyway... What do you @masterleinad @davydden think?

davydden replied:
i think there is no need for a separate communicator. IMHO it should be consistent with the one used in `p::Tria`. From this perspective i very much like these functions are parallel/non-parallel is hidden in implementation.

masterleinad replied:
@kronbichler I totally agree that setting `mpi_communicator` can be internal to `MatrixFree`.

@bangerth I don't quite see the point to do the `dynamic_cast` twice instead of having a `const` variable if this is what you are suggesting.

kronbichler replied:
See #3607.

bangerth replied:
@masterleinad -- I was thinking of the idiom
```
  if (const parallel::Triangulation<dim,spacedim> *parallel_tria =
      dynamic_cast<const parallel::Triangulation<dim,spacedim>*>(&dof.get_tria()))
    additional_data.mpi_communicator = parallel_tria->get_communicator();
```
I should have been clearer given that this is a very useful, but not widely known idiom.

The point may be moot given the discussion above about not allowing a separate communicator to be set.

davydden replied:
i did not know that you can do this, nice!

masterleinad replied:
Yes, that looks nice!

bangerth replied:
I think it's an under-appreciated idiom.

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1160,6 +1293,141 @@ namespace VectorTools
+    mass_matrix.vmult_add(rhs, inhomogeneities);
+
+    //now invert the matrix
+    ReductionControl     control(5.*rhs.size(), 0., 1e-12, false, false);
```
same issue with the number of iterations.

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -1160,6 +1293,141 @@ namespace VectorTools
+
+
+  template <int dim, typename VectorType>
+  void project_matrix_free (const Mapping<dim>   &mapping,
```
This function is not declared in the `.h` file. Should it be, i.e., should it be a public and documented interface? If not, I'm in favor of moving it into the `internal` namespace.

davydden replied:
i would say not, it should be internal

bangerth commented on source/numerics/vector_tools_project.inst.in:
```diff
@@ -20,7 +20,7 @@ for (VEC : SERIAL_VECTORS ; deal_II_dimension : DIMENSIONS; deal_II_space_dimens
     namespace VectorTools \{
```
same issue with file permissions

bangerth commented on tests/fe/fe_values_view_30.cc:
```diff
@@ -109,7 +109,7 @@ void test (const Triangulation<dim> &tr,
   MappingQGeneric<dim> mapping(1);
 
-  const QGauss<dim> quadrature(2);
+  const QGauss<dim> quadrature(fe.degree+2);
```
What was the reason for this change? Did you just want to check more data, or did it actually uncover an issue?

masterleinad replied:
Actually, I just wanted to have the same output for the new and the old code. Since, `project_generic` uses `QGauss<dim> quadrature(fe_degree+2)` this implies to use the same quadrature for the right-hand side.

bangerth replied:
Hm, I may have missed this -- `fe_values_view_30` calls the `project()` function, which now produces different output than it did before? Or do I misunderstand your comment?

masterleinad replied:
The changed test passes for the old implementation as well as for the new one. So yes, changing the test changes the output.

The old implementation uses the given `Quadrature` for assembling both the `system_matrix` and the `rhs`
while the new implementation always uses `QGauss(fe_degree+2)` for the `system_matrix` and the given `Quadrature` for the `rhs`.

bangerth commented on tests/numerics/project_parallel_common.h:
```diff
@@ -0,0 +1,358 @@
+
+
+// common framework to check whether an element of polynomial order p can
+// represent functions of order q
```
Is this really what the test does? It seems about projection. Maybe expand the comment.

bangerth commented:
It's too late now for this patch, but if I may suggest something for future cases: This PR changes many unrelated things:
* It adds new functions
* It also adds a `VectorType` template argument to a bunch of functions
* It changes a few tests that seem unrelated by changing the quadrature order
* It inlines the name of the output file in a few tests (e.g., `kelly_crash_02`)

It makes patches awkward to review if there are different things. It also makes them large, which I'm sure has something to do with the fact that this patch was slow to get reviewed (my apologies for that as well). Going forward, here is what I often do in situations like this:
* I commit things in small batches, and often add new changes to existing commits via interactive rebase
* I end up with a bunch of commits on a branch, where each commit is reasonably self-contained
* I create a new branch off of master where I cherry-pick a commit that is ready for review, regardless of whether the other parts of the functionality are already ready. An example could have been the output file name changes for the testsuite
* I submit this one commit for peer review
* Since that commit is still in my feature branch, I can continue working on that branch and don't need to worry about the just committed pull request
* When the pull request is merged, I just do a rebase. The commit that I had both on my current branch and the branch from which I created the pull request magically disappears in the rebase (because it's already in the master upon which the branch is now based), and the diff of my current branch has become magically smaller.

In fact, I think you could pretty easily still do that with the commits you have here.

Anyway, just an idea going forward to make patch review a bit easier (and faster -- after all, we've all let you hang out there with this patch for quite a while already!).

davydden commented:
@masterleinad : to follow-up on @bangerth 's suggestion, maybe you can extract first two commits from this PR and the last two into separate PRs, while we continue discussing the project-only functions here.

masterleinad commented:
@bangerth Thanks for the suggestion! Actually, I feel a bit uncomfortable that the state in this PR is not compiling until at least  #3610 is merged, but let's see.

davydden commented:
@masterleinad i think the point was NOT to remove old commits from this PR, in other words keep it with all those 5 commits you had. Then after #3610 is merged, you would be able to re-based on top of `master` and won't loose anything. Meanwhile your branch would have also been fully functional, which it is not now as you got rid of those commits.

bangerth commented:
Like @davydden says -- git has this nice feature where if you have a history like
```
  A -> B -> C
```
on a branch (where A is where master is), and you extract B onto a separate branch, i.e.,
```
  A -> B -> C
    \
      B'
```
where B and B' have the same content, then if that side branch gets merged to yield
```
      B -> C
    /
  A ---> X
    \    /
      B'
```
where master is now on X, then if you rebase the `B->C` branch to current master (X), you'll get this:
```
               C
              /
  A ---> X
    \    /
      B'
```
In other words, it magically dropped B from your branch because the equivalent of B (namely B') is already in the history. You don't actually have to pull B from out of your branch by hand.

masterleinad commented:
Sure, I thought you wanted to have just the changes due to `VectorTools::project` here.

davydden commented:
> Sure, I thought you wanted to have just the changes due to VectorTools::project here.

eventually that is going to be the case once other PRs are merged, but there was no need to cut-off those extra commits at this point in time.

masterleinad commented:
This should address all requested changes.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,267 @@ namespace VectorTools
 
 
+    /*
+     * Generic implementation for project using a MatrixFree implementation
```
looking at the usage, i think it should be called `project_matrix_free`.

davydden replied:
please remove `Generic`. Also for->of.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+
+
+    template <int dim, typename VectorType>
+    void project_matrix_free (const Mapping<dim>            &mapping,
```
this name is mis-leading here as you choose between matrix-free and old implementation. Since this is called only for dim==spacedim, maybe call it `project_dim_eq_spacedim`

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+      * and call project_generic with the appropriate template arguments.
+      */
+    template <int components, int dim, typename VectorType, int spacedim>
+    void project_parallel_helper (const Mapping<dim, spacedim>                              &mapping,
```
i would rename this one to `project_matrix_free_degree` or something like that so that the name indicates that you have a switch inside based on degrees.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+    // Helper interface for the matrix-free implementation of project().
+    // Used to determine the number of components.
+    template <int dim, typename VectorType, int spacedim>
+    void project_parallel (const Mapping<dim, spacedim>           &mapping,
```
i would also rename this one to something like `project_matrix_free_component`, again to better indicate that you switch based on the component.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+          else
+            {
+              const FiniteElement<dim> *fe_ptr = &dof.get_fe().base_element(0);
+              if (fe_ptr->n_components() != 1)
```
`<= 4` here.

masterleinad replied:
No, this comes from `shape_info.templates.h` as well. In particular, for `FESystem<dim>(FE_Q<dim>,dim)` the base element is still `FE_Q` which has only one component.

If this were the number of components of the `FiniteElement`, the `project_parallel_q_*` tests would certainly fail.

davydden replied:
right, you are testing 0th base element, not the whole thing.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+          break;
+
+        default:
+          Assert(false, ExcNotImplemented());
```
throw an `ExcInternalError()` here because you check already components in the function with `bool use_matrix_free`.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+          break;
+
+        default:
+          Assert(false, ExcNotImplemented());
```
throw an ExcInternalError() here because you already check supported degrees in the function with `bool use_matrix_free.`

masterleinad replied:
Actually, I neither check for the number of components nor for the degree of the `FiniteElement` with `use_matrix_free`, but I see the point in changing this.

davydden commented:
Apart from minor suggestions, it looks good to me! :+1:

masterleinad commented:
@davydden Great! I mainly followed your suggestions for renaming functions, but got away with `project_matrix_free->project` by swapping template arguments.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,268 @@ namespace VectorTools
+
+    /**
+      * Helper interface. After figuring out the number of components
+      * in project_parallel, we determine the degree of the FiniteElement
```
project_parallel-> project_matrix_free_component

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,268 @@ namespace VectorTools
+    }
+
+
+
```
add a few lines of what this function does.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -891,6 +895,268 @@ namespace VectorTools
+      else
+        {
+          // Find out if the FiniteElement is supported
+          // This is copied from matrix_free/shape_info.templates.h
```
copied -> based on (because you add extra checks in this part of the code).

masterleinad commented:
@davydden Here you go!

kronbichler commented on tests/numerics/project_parallel_common.h:
```diff
@@ -0,0 +1,359 @@
+                                                mpi_communicator);
+  TrilinosWrappers::MPI::Vector projection (locally_owned_dofs,
+                                            locally_relevant_dofs,
+                                            mpi_communicator);
```
Is there a deeper reason to use Trilinos vectors here? You could also use `LinearAlgebra::distributed::Vector`, which would have the advantage that the test runs also without Trilinos.

kronbichler commented on tests/numerics/project_parallel_q_01.with_trilinos=true.mpirun=1.output:
```diff
@@ -0,0 +1,23 @@
+
```
I think you need to append `.with_mpi=true` here as well.

kronbichler commented on tests/numerics/project_parallel_q_01.with_trilinos=true.mpirun=4.output:
```diff
@@ -0,0 +1,23 @@
+
```
Same here

kronbichler commented on tests/numerics/project_parallel_q_03.with_trilinos=true.mpirun=1.output:
```diff
@@ -0,0 +1,78 @@
+
```
I think that it would be enough to only run the test `project_parallel_q_01` with both 1 and 4 processors and the `02` through `05` tests only with 4 processors. Even though we like to have as many test as possible, running the exact same thing 20 times (debug, release, 5 test variants that only differ in some aspects of the mesh) seems a bit too much.
